doc/sfp-cmis/cmis-init.md

```diff
@@ -0,0 +1,301 @@
+- If the transceiver is a CMIS module, the **show interfaces transceiver eeprom**
+should display the advertised applications of the transceiver.
+- If the transceiver is a CMIS module, the **sfputil** should be capable of detecting the errors of the CMIS application initialization.
+
+# Design
+
+## sonic-platform-daemons/sonic-xcvrd (modified)
+
+The transceiver daemon will be enhanced as below:
+
+- Post the CMIS application advertisement into the **STATE_DB**  
+  In the case of CMIS transceiver, the **application_advertisement** of **TRANSCEIVER_INFO**
+is updated from plain test into json dictionary format:  
+
+**Original:**  
+```
+  "TRANSCEIVER_INFO|Ethernet0": {
+    "type": "hash",
+    "value": {
+      "application_advertisement": "400GAUI-8 C2M (Annex 120E) - 400GBASE-DR4 (Cl 124)\n        100GAUI-2 C2M (Annex 135G) - 100G-FR/100GBASE-FR1 (Cl 140)",
+      ...... omitted ......
+    }
+  },
+```
+
+**Modified:**  
+```
+  "TRANSCEIVER_INFO|Ethernet0": {
+    "type": "hash",
+    "value": {
+      "application_advertisement": "{1: {'ap_sel_code_id': 1, 'host_electrical_interface_id': '400GAUI-8 C2M (Annex 120E)', 'module_media_interface_id': '400GBASE-DR4 (Cl 124)', 'host_lane_count': 8, 'media_lane_count': 4, 'host_lane_assignment_options': 1, 'media_lane_assignment_options': None}, 2: {'ap_sel_code_id': 2, 'host_electrical_interface_id': '100GAUI-2 C2M (Annex 135G)', 'module_media_interface_id': '100G-FR/100GBASE-FR1 (Cl 140)', 'host_lane_count': 2, 'media_lane_count': 1, 'host_lane_assignment_options': 85, 'media_lane_assignment_options': None}}",
+      ...... omitted ......
+    }
+  },
+  ...... omitted ......
+```  
+
+- Add **CMIS_MEDIA_SETTINGS** to the media_settings.json for custom CMIS controls
+
+**Format:**
+```
+{
+    "CMIS_MEDIA_SETTINGS":{
+        "<VENDOR_NAME>#<PART_NUMBER>": {
+            "application_selection": true|false
+        }
+    }
+}
+```
+
+**Example:**
+```
+{
+    "CMIS_MEDIA_SETTINGS":{
+        "AVAGO#AFCT-93DRPHZ-AZ2": {
+            "application_selection": true
+        },
+        "Amphenol#NDYYYF-0001": {
+            "application_selection": false
+        },
+        "DELL EMC#6MGDY": {
+            "application_selection": true
+        },
+        "DELL EMC#DH11M": {
+            "application_selection": false
+        }
+    }
+}
+```
+  As of now, only **application_selection** is supported, and it defaults to **True** if unspecified
+
+- Add a CMIS manager class to handle the application initializations in parallel  
+
+  ![](images/001.png)
+
+- Add **PortChangeEvent.PORT_SET** and **PortChangeEvent.PORT_DEL** events to **xcvrd_utilities/port_mapping.py**  
+  - PortChangeEvent.PORT_SET  
+    Invoke the event callbacks upon **SET** operations, it includes all the port config updates,
+not just the port add/remove events.
+  - PortChangeEvent.PORT_DEL  
+    Invoke the event callbacks upon **DEL** operations.
+
+- As of now, the number of CmisManagerTask is hard coded to 4  
```

Comment:
I think we should take a strong look at xcvrd and consider creating a separate processing thread per xcvr. This allows for all slow i2c operations (DOM update, etc) to be done in a non-blocking manner.

Regardless, 4 is certainly not enough. If we assume 5 seconds per application selection and 32 modules, it would still take ~40 seconds to initialize the modules when it could actually be done in ~5s.

Reply:
Yes, agree, the CMIS transactions could take several seconds to finish (while it's possible to even scale up to minutes as per the CMIS SPEC, such modules are unlikely to be deployed in the SONIC, as of now)